Register post terms variations on taxonomy registration

gaambo opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
This solves the issue from #52569 where custom taxonomies registered after `init` priority 10 are not taken into account for the post-terms blocks variations.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
See https://github.com/WordPress/gutenberg/issues/52569 - server side function runs too early to make sure it gets all custom taxonomies/post types by plugins, themes etc..

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Similar to #54801 (which does the same thing for the navigation link block)
Variations for all taxonomies registered until `register_block_core_post_terms` (at init priority 10) are registered directly in here. For all taxonomies registered after that, actions are added to `registered_taxonomy`, which will register variations during taxonomy registration.

We can't use the second approach for all taxonomies, because some of them (eg builtin ones) are registered before the post terms block is registered, therefore variations can't yet be added.

I'm directly adding the variation on the WP_Block_Type object from WP_Block_Type_Registry, because there's no API for registering variations in PHP (yet).

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
### Manual testing
#### Taxonomy
1. Register a custom taxonomy ([example code from developer.wordpress.org](https://developer.wordpress.org/plugins/taxonomies/working-with-custom-taxonomies/#step-2-creating-a-new-plugin)) 
```php
add_action('init', function() {
	 $args   = array(
		 'labels'            => array(
		     'name'         => __( 'Courses' ),
	         ),
	 );
	 register_taxonomy( 'course', [ 'post' ], $args );
});
```
Important: Set the name/label or the block variation is called Post Link.
Important: change add_action priority to something higher than 10 (eg 11) - when using something lower it already works in trunk.
3. Go into site- or post-editor (for the post type post)
4. try to search for a block named like the custom taxonomy (eg "Courses").
5. Block is available

### Automated testing

Run `npm run test:unit:php:base -- --filter Block_Post_Terms_Variations_Test`

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

gaambo committed:
register variations on taxonomy registration

gaambo committed:
add unit test for variations

gaambo commented:
@draganescu Could you please have a look at this as well? It's pretty much the same as #54801, but for the post-terms block.

gaambo committed:
fix public check for taxonomies

gaambo committed:
add unit test for private taxonomies

gaambo committed:
handle taxonomy unregistration

draganescu reviewed (approved):
This looks good to me, I found myself wondering if unregistering before the block registration needs anything but it does not since once unregistered the tyeps won't exist for the block registration to build variations on them.

👏🏻

draganescu commented:
Depending on what you get on the previous core ticket (if we need to port this in any way) this PR is subject to the same conclusions.

gaambo commented:
@desrosj since we've got #54801 and #56100 merged (for the navigation-link block), is there any chance this could get merged as well? 
I probably have to update this PR as well, to work with the changes about the `WP_Block_Type::variations` property made in 2c247e2ae62b5d61dcdf5c634c13c7ccf0731b0d and #56952 as well.

gaambo committed:
fix WP_Block_Type::$variations changes

gaambo commented:
Updated the PR to be similar to the changes in #56952 and work with the new private `variations` property of WP_Block_Type and `variation_callback`. I also renamed the functions so the post-terms block PHP is very similar to the navigation-link ones.

gaambo committed:
fix code style of comments

draganescu commented:
@gaambo what happened with this one? I see you closed it but what's the replacement?

gaambo commented:
@draganescu 
I've closed it, because the original implementation (via `registered_post_type` and `registered_taxonomy`) was also used in #54801 for the navigation link block and that lead to timeouts during PHPUnit tests (see https://github.com/WordPress/gutenberg/pull/54801#discussion_r1469687143). 
For the navigation link block, there's currently a workaround by unregistering this hooks when running PHPUnit tests (see https://github.com/WordPress/wordpress-develop/pull/5922#issuecomment-1915557316) and I've made a follow-up PR which uses the new `get_block_type_variations` hook instead: #58389

I want to see which way it goes for the navigation link block and when that's settled I will do the same for the post-terms block, since it's pretty similar and quick to do then.

gaambo commented:
FYI: this was already fixed in https://github.com/WordPress/gutenberg/commit/2c247e2ae62b5d61dcdf5c634c13c7ccf0731b0d via the new `variation_callback` argument when registering the block.
